src/shared/dm-util.c

```diff
@@ -43,3 +43,39 @@ int dm_deferred_remove(const char *name) {
         return 0;
 #endif
 }
+
+int dm_deferred_remove_cancel(const char *name) {
+        _cleanup_close_ int fd = -1;
+        _cleanup_free_ struct dm_ioctl *dm = NULL;
+        const uint32_t dm_size = sizeof(struct dm_ioctl) + sizeof(struct dm_target_msg) + strlen("@cancel_deferred_remove") + 1;
+        struct dm_target_msg *msg = NULL;
+
+        assert(name);
+
+        if (strlen(name) >= DM_NAME_LEN)
+                return -ENODEV; /* A device with a name longer than this cannot possibly exist */
+
+        dm = malloc0(dm_size);
```

Comment:
this is fixed size. Hence maybe just allocate this on the stack? And then you can even use structured initialization...

```
struct {
        struct dm_ioctl dm_ioctl;
        struct dm_target_msg dm_target_msg;
        char text[STRLEN("@canc…") + 1];
} message = {
        .dm_ioctl = {
                .version[0] = DM_VERSION_MAJOR;
…
```

Reply:
Looks neat - do I need to worry about packing/padding? Perhaps set data_size as sizeof(message) and data_start as &message.dm_target_msg ? Can never remember what the standard says about these details

Reply:
Applied and pushed